### DISCUSSION TITLE
Role-Based Routing

### DISCUSSION THREAD

**Discussion by adhenrique** [vercel/next.js#23041] [Help]
Hello friends.

I'm new to NextJs and I'm implementing an admin area with some roles: admin, manager and employee.
In the pages folder I would have 2 subfolders to share the UI: `admin/` and `employee/` - since the user ui with role `employee` is completely different.

I am looking to implement some verification at the UI level, suddenly using SSR, to ensure that the user has the necessary role to view a certain page. In other words:
`admin/users/all.js` -> only users with role `admin` can view this page
If a user with a role of `employee`, for example, tries to request that page, I want to redirect him to his home: `employee/index.js`

How difficult is it to do this? I believe that I am mixing the subjects regarding getServerSideProps and getInitialProps in the custom `_app.js`. I found very few approaches about it...

Note: my API already checks whether the user is allowed to view the data for a particular page, but this is still an API-level check. Again, I am thinking of something at the UI level.

Can someone help me with any approach?

**Comment by ErnestBrandi** [comment]
Hi, let suppose you have a `user` object in those pages, then you can do : 
```
React.useEffect(() => {
   if(user.role !== "admin") router.push('/whaterver-page');
}, [user]);
```
https://reactjs.org/docs/hooks-reference.html#useeffect
https://nextjs.org/docs/api-reference/next/router

**Reply by adhenrique** [reply]
Thanks @ErnestBrandi 

That is one way. But can I use this way inside `_app.js`? That way I wouldn't need to repeat this code within each page - unless I write a custom hook... but I would still repeat it on every page...

**Reply by ErnestBrandi** [reply]
Well React is all about composing so including one line of code in order to get an certain result seems fine to me. You cannot really put this in `_app` because that would mean the whole website would become inaccessible for a non-admin logged user.

**Comment by metin1** [comment]
Hi, 
I was looking for role based auth yesterday. I code like this:

authContext.js:
````
import React, {
  createContext,
  useContext,
  useMemo,
  useEffect,
} from "react";
import { useCurrentUser } from "@/hooks/index";
import { useRouter } from "next/router";
import UnAuthPage from "@/components/Layout/pages/unauth";

const AuthStateContext = createContext();

export function useAuth() {
  return useContext(AuthStateContext);
}

function AuthProvider({ children }) {
  const [user, { mutate }, { isLoading }] = useCurrentUser();
  const router = useRouter();
  let role = user?.role;
  let allowed = true;

  if (router.pathname.startsWith("/user") && role !== "user") {
    allowed = false;
  }
  if (router.pathname.startsWith("/teacher") && role !== "teacher") {
    allowed = false;
  }
  if (router.pathname.startsWith("/applicant") && role !== "applicant") {
    allowed = false;
  }
  if (router.pathname.startsWith("/admin") && role !== "admin") {
    allowed = false;
  }

  useEffect(() => {
    role = user?.role;
    console.log(`LL: AuthProvider -> role`, role);
    if (!isLoading) {
      if (router.pathname.startsWith("/user") && role !== "user") {
        if (role) {
          console.log(router.pathname.replace("user", role));
          router.push(router.pathname.replace("user", role));
        } else {
          // router.push("/");
        }
      }
      if (router.pathname.startsWith("/teacher") && role !== "teacher") {
        if (role) {
          console.log(router.pathname.replace("teacher", role));
          router.push(router.pathname.replace("teacher", role));
        } else {
          router.push("/");
        }
      }
      if (router.pathname.startsWith("/applicant") && role !== "applicant") {
        if (role) {
          console.log(router.pathname.replace("applicant", role));
          router.push(router.pathname.replace("applicant", role));
        } else {
          router.push("/");
        }
      }
      if (router.pathname.startsWith("/admin") && role !== "admin") {
        if (role) {
          console.log(router.pathname.replace("admin", role));
          router.push(router.pathname.replace("admin", role));
        } else {
          router.push("/");
        }
      }
    }
  }, [user, router]);

  const contextValue = useMemo(() => [allowed, role], [allowed, role]);
  const ComponentToRender = allowed ? children : UnAuthPage;

  return (
    <AuthStateContext.Provider value={contextValue}>
      {ComponentToRender}
    </AuthStateContext.Provider>
  );
}

export { AuthProvider };
```

and 
_app.js

```

import { AuthProvider } from "@/components/providers/authContext";

const App = ({ Component, pageProps }) => {
return (
    <AuthProvider>
      <Component {...pageProps} />
    </AuthProvider>
)
}

**Reply by metin1** [reply]
This is works for me. I don't need to code every page. Only I have only a problem with history back. If the user tried to unauth routes then this code forwards to the correct roled route. But after that process, the user wants to go back show an empty screen with auth address. If I find any solution I can add it here. Thanks

**Reply by adhenrique** [reply]
@metin1 thank you very much. With this piece of code I can already see a light at the end of tunnel. I will try to create something on top of these proposals and I will return soon.

**Reply by windanetto** [reply]
hi, @metin1 I try your code, but I got error in many renders. can you help me to fix it? or anybody?

here authContext.js:

`import router, { useRouter } from "next/router";
import React, {
  createContext,
  useContext,
  useEffect,
  useMemo,
  useState,
} from "react";
import { pageNotFound } from "../../pages/404.js";

const AuthStateContext = createContext();

export default function useAuth() {
  return useContext(AuthStateContext);
}

function AuthProvider({ children }) {

  const router = useRouter();
  const [user, setUser] = useState();
  const [roleId, setRoleId] = useState();
  const [allowed, setAllowed] = useState(true);
  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    setIsLoading(true);
    let loginData = null;
    if (typeof window !== "undefined") {
      const loginDataString = localStorage.getItem("loginData");
      loginData = loginDataString ? JSON.parse(loginDataString) : null;
      if (loginData) {
        setUser(loginData);
        setRoleId(loginData.userRoles[0].roleId);
        setIsLoading(false);
      }
    }
  }, []);

  if (router.pathname.startsWith("/administrator") && roleId === 7) {
    setAllowed(false);
  }

  useEffect(() => {
  
    if (!isLoading) {
      if (router.pathname.startsWith("/administrator") && roleId === 7) {
        if (roleId) {
          console.log(router.pathname.replace("/", roleId));
        } else {
          router.push("/");
          console.log("masuk else >>>");
        }
      }
    }
  }, [user, router]);

  const contextValue = useMemo(() => [allowed, roleId], [allowed, roleId]);
  const ComponentToRender = allowed ? children : pageNotFound;

  return (
    <AuthStateContext.Provider value={contextValue}>
      {ComponentToRender}
    </AuthStateContext.Provider>
  );
}

export { AuthProvider };
`
this is my _apps.js:
`import React from "react";
import ReactDOM from "react-dom";
import App from "next/app";
import Head from "next/head";
import Router from "next/router";
import PageChange from "components/PageChange/PageChange.js";
import "@fortawesome/fontawesome-free/css/all.min.css";
import "styles/tailwind.css";
import "styles/global.css";
import "rsuite/dist/rsuite.min.css";
import { AuthProvider } from "../components/Providers/AuthContext";


export default class MyApp extends App {
  componentDidMount() {
    let loginData = null;
    if (typeof window !== "undefined") {
      const loginDataString = localStorage.getItem("loginData");
      loginData = loginDataString ? JSON.parse(loginDataString) : null;
      if (!loginData) {
        Router.push("/sign/in");
      }
    }
    let comment = document.createComment(`

=========================================================
* Notus NextJS - v1.1.0 based on Tailwind Starter Kit by Creative Tim
=========================================================

`);
    document.insertBefore(comment, document.documentElement);
  }
  static async getInitialProps({ Component, router, ctx }) {
    let pageProps = {};

    if (Component.getInitialProps) {
      pageProps = await Component.getInitialProps(ctx);
    }

    return { pageProps };
  }
  render() {
    const { Component, pageProps } = this.props;

    const Layout = Component.layout || (({ children }) => <>{children}</>);

    return (
      <React.Fragment>
        <Head>
          <meta
            name="viewport"
            content="width=device-width, initial-scale=1, shrink-to-fit=no"
          />
          <title>Admin Dashboard</title>
          <script src="https://maps.googleapis.com/maps/api/js?key=YOUR_KEY_HERE"></script>
        </Head>
        <Layout>
          <AuthProvider>
            <Component {...pageProps} />
          </AuthProvider>
        </Layout>
      </React.Fragment>
    );
  }
}
`

**Comment by revskill10** [comment]
One simple solution, is you provide something like `Component.allowed_roles = ["manager"]` for each page.

Then in `_app.js`, just check the current user role with `Component.allowed_roles`

**Reply by tymoxx** [reply]
@revskill10, could you elaborate more on what you would write in `_app.js`?

**Comment by sujin-sreekumaran** [comment]
Dosn't this much of code slows the render process??
